pkg/producer/kafka.go

```diff
@@ -164,14 +179,40 @@ func (kp *KafkaProducer) SendFlowMessage(msg protoreflect.Message, kafkaDelimitM
 
 }
 
-// Publish takes in a message channel as input and converts all the messages on
+// PublishIPFIXMessages takes in a message channel as input and converts all the messages on
 // the message channel to flow messages in proto schema. This function exits when
 // the input message channel is closed.
-func (kp *KafkaProducer) Publish(msgCh chan *entities.Message) {
+func (kp *KafkaProducer) PublishIPFIXMessages(msgCh chan *entities.Message) {
 	for msg := range msgCh {
 		flowMsgs := kp.protoSchemaConvertor.ConvertIPFIXMsgToFlowMsgs(msg)
 		for _, flowMsg := range flowMsgs {
 			kp.SendFlowMessage(flowMsg, true)
 		}
 	}
 }
+
+func (kp *KafkaProducer) PublishRecord(record entities.Record) {
+	flowMsg := kp.protoSchemaConvertor.ConvertIPFIXRecordToFlowMsg(record)
+	kp.SendFlowMessage(flowMsg, false)
+}
+
+func (kp *KafkaProducer) Close() {
+	if kp.producer == nil {
+		return
+	}
+
+	var wg sync.WaitGroup
+	kp.producer.AsyncClose()
+
+	wg.Add(1)
+	go func() {
+		close(kp.closeCh)
+		if kp.input.KafkaLogSuccesses {
+			for range kp.producer.Successes() {
+				klog.Error("Unexpected message on Successes()")
+			}
```

Comment:
under what circumstances would this case happen?

Reply:
The producer is async producer. When `sendFlowMessage` and `Close` called simultaneously, this may happen.